internal/rbd/rbd_attach.go

```diff
@@ -239,6 +241,9 @@ func appendDeviceTypeAndOptions(cmdArgs []string, isNbd, isThick bool, userOptio
 		if !strings.Contains(userOptions, setNbdReattach) {
 			cmdArgs = append(cmdArgs, "--options", fmt.Sprintf("%s=%d", setNbdReattach, defaultNbdReAttachTimeout))
 		}
+		if !strings.Contains(userOptions, setNbdIOTimeout) {
+			cmdArgs = append(cmdArgs, "--options", fmt.Sprintf("%s=%d", setNbdIOTimeout, defaultNbdIOTimeout))
```

Comment:
Question:
@pkalever will there be any impact on pod graceful termination and force termination because of this one?

Reply:
@Madhu-1 Nope this shouldn't hurt the termination sequence of the pod lifecycle. This timeout is supplied to the nbd driver via rbd-nbd process, It means how long the cmds should wait before timing out (like: when the rbd-nbd process is down, wait for it to serve IO)


But it's good to get some confirmation from @idryomov about the values we are using i.e reattach-timeout=300 and io-timeout=$reattach-timeout + 30, has any further objections.

cc: @lxbsz

Reply:
This seems wrong to me.  I haven't tested but I think this would cause in-flight requests to not be requeued when rbd-nbd is respawned.  See https://github.com/torvalds/linux/commit/2c272542baee2c3b9e8e3a260db81227ccefe8b5.

Reply:
I have tested the different settings of io-timeout values like greater than, less than, equal to reattach timeout value keeping reattach timeout at 300 fixed, and honestly with the latest nbd driver available at the master  I had not seen any difference in the io timeout/fail behavior.

Here are the observations captured: https://hackmd.io/@FFELd2RJT76Uqp-3M-e4zw/Sys8bLfeY

I'm planning to test the behaviour on the minikube kernel version to understand the behaviour between the kernels.

---

I hope now you might sense the need for https://github.com/ceph/ceph/pull/42609

Thanks!

CC: @trociny @lxbsz

Reply:
On the contrary, so far I don't see the need for https://github.com/ceph/ceph/pull/42609 at all.

I don't understand what you were testing though.  If you kill rbd-nbd with `reattach-timeout` set, eventually in-flight I/O would be failed no matter what the value of `io-timeout` is.  I think the test should have included `rbd device attach` at e.g. 3 minute mark with the expectation that nothing gets timed out, the filesystem stays read-write and dd continues to execute after reattach.  I would expect `io-timeout=30` to be disqualified immediately and with enough test rounds you should be able to see the difference between `io-timeout=300`/`io-timeout=330` and `io-timeout=0`.

Reply:
Nit: get dmesg dumps with raw jiffie timestamps (i.e. none of that `Aug12 06:52` and `+0.004259` stuff).  Given that we are investigating 30 second timeouts, a timestamp such as `Aug12 06:52` that doesn't include seconds at all is pretty useless.

Reply:
> On the contrary, so far I don't see the need for [ceph/ceph#42609](https://github.com/ceph/ceph/pull/42609) at all.

We don't do -osync writes from mount, which means we may end up with some IO pending at filesystem/block layers while the rbd-nbd goes detached and we shouldn't expect the pending IO in the filesystem/block layers to be handled as part of SIGTERM handler of rbd-nbd.

1. The quiesce will ensure that the data will be pushed to the rbd-nbd layer.
2. It will also block/wait the IO on the mount while the rbd-nbd is offline.
3. Also just in case if the process is not attached in the reattachtimeout, the copy is ensured to be consistent (corruption free)
4. Snapshots can be taken in a consistent way even when the rbd-nbd process is offline.

> 
> I don't understand what you were testing though. If you kill rbd-nbd with `reattach-timeout` set, eventually in-flight I/O would be failed no matter what the value of `io-timeout` is. I think the test should have included `rbd device attach` at e.g. 3 minute mark with the expectation that nothing gets timed out, the filesystem stays read-write and dd continues to execute after reattach.

I was testing a case where the attach doesn't happen in the reattach-timeout and wanted to note how IO behaves in various io-timeouts used.

I expect the process attach to handle the IO which is not timed out yet, but what about the IO which is timed out already (30 secs is too low, I have not seen them timing out in 30secs though)? We cannot expect the process to be back in 30 seconds and cope up with the pending IO. And we are uncertain how to handle the read-only mounts which are perhaps happening after the IO timeouts. Certainly, we want to avoid this situation.

> I would expect `io-timeout=30` to be disqualified immediately and with enough test rounds you should be able to see the difference between `io-timeout=300`/`io-timeout=330` and `io-timeout=0`.

I will test more with the dmesgs timestamps suggestions you mentioned. Also include tests with attach next time.

Thanks @idryomov.

Reply:
> > On the contrary, so far I don't see the need for [ceph/ceph#42609](https://github.com/ceph/ceph/pull/42609) at all.
> 
> We don't do -osync writes from mount, which means we may end up with some IO pending at filesystem/block layers while the rbd-nbd goes detached and we shouldn't expect the pending IO in the filesystem/block layers to be handled as part of SIGTERM handler of rbd-nbd.

Right, that I/O would get processed after rbd-nbd is reattached.

> 
>     1. The quiesce will ensure that the data will be pushed to the rbd-nbd layer.
> 
>     2. It will also block/wait the IO on the mount while the rbd-nbd is offline.
> 
>     3. Also just in case if the process is not attached in the reattachtimeout, the copy is ensured to be consistent (corruption free)
> 
>     4. Snapshots can be taken in a consistent way even when the rbd-nbd process is offline.

This is true but I don't see much value in smoothing out the "reattach didn't happen" case, especially if it is achieved by allowing more or less arbitrary code execution within the signal handler.  The only reason we are bothering with all of this (`rbd device attach`, `io-timeout`, etc) is to make reattach not just possible but also reliable.  It should "just" work and if something goes wrong, it is not that different from an abrupt termination (e.g. if rbd-nbd gets a SIGKILL from k8s because it doesn't react to SIGTERM in time or if rbd-nbd actually crashes due to a bug in librbd or the daemon itself).

Is there a reason for treating the "reattach didn't happen" case specially?

Reply:
Currently this is the most important requirement

> 
> 2. It will also block/wait the IO on the mount while the rbd-nbd is offline.
> 

When the re-attach is not fast enough, or the filesystem decides do write to the block device while the rbd-nbd process is unavailable, the (ext4) filesystem detects a problem with the block device, and decides to become read-only. There is no reasonable way for a Pod to recover from that.

Before the rbd-nbd process does a detach, it would be best to call `fsfreeze` and block I/O from going to the block-device. Upon attach an unfreeze should be done. It is ideal if `rbd-nbd` support this, but we can consider an alternative where Ceph-CSI tries to handle it.

Reply:
Well honestly, I do not see a good reason why rbd-nbd shouldn't do it, the way like snapshots.

That said, ceph-csi cannot handle it, because although the rbd-nbd process and the application mount resides on the same node they both run in two different pods, so ceph-csi cannot coordinate b/w freeze and detach. Since the rbd-nbd receives the signal in any case, it can very well handle it.

If we have to do it from the `preStop` hook, at scale, it will be a lot of serializing of all the freezes tasks per PV. Since there will be one rbd-nbd process per volume, the fsfreeze will be run parallelly in case rbd-nbd can handle it, that way we will not hit the termination grace period of Kubernetes.

Reply:
> When the re-attach is not fast enough, or the filesystem decides do write to the block device while the rbd-nbd process is unavailable, the (ext4) filesystem detects a problem with the block device, and decides to become read-only. There is no reasonable way for a Pod to recover from that.

Even with `io-timeout` set to 0?  I don't think ext4 detects anything by itself.  IIUC the problem was that that we weren't setting `io-timeout` and the block layer was returning an error after 30 seconds.  This PR should resolve that.

As for what happens if reattach doesn't occur within `reattach-timeout`, if the filesystem is frozen, wouldn't the user application just hang in D state?  How would you recover from that?

Reply:
> This is true but I don't see much value in smoothing out the "reattach didn't happen" case, especially if it is achieved by allowing more or less arbitrary code execution within the signal handler. The only reason we are bothering with all of this (`rbd device attach`, `io-timeout`, etc) is to make reattach not just possible but also reliable. It should "just" work and if something goes wrong, it is not that different from an abrupt termination (e.g. if rbd-nbd gets a SIGKILL from k8s because it doesn't react to SIGTERM in time or if rbd-nbd actually crashes due to a bug in librbd or the daemon itself).

1. Once we have the rbd-nbd doing the freeze, the idea is to improve the termination sequence of the rbd nodeplugin pod by adding the preHooks such a way that Kubernetes waits for all the rbd-nbd processes on the pod before doing any ungrateful shutdown, yes I remember some of ceph-csi team conversations and it really make sense!

2. Bug in librbd or daemon is not acceptable and should be resolved once we root cause it :-)

> 
> Is there a reason for treating the "reattach didn't happen" case specially?

Because, that is the worst that can happen in a normal work flow, as far as we can think.

Thanks!

Reply:
@idryomov when io-timeout is set to 0, I do not see any difference that is getting affected.


```
[root@linux-vm1 build]# ./bin/rbd-nbd map rbd-pool/image0 --try-netlink --reattach-timeout 300 --io-timeout 0
/dev/nbd0

[root@linux-vm1 build]# cat /sys/block/nbd0/queue/io_timeout 
30000
```

This is why I treat it to behave the same as default 30 seconds.

Reply:
> * Once we have the rbd-nbd doing the freeze, the idea is to improve the termination sequence of the rbd nodeplugin pod by adding the preHooks such a way that Kubernetes waits for all the rbd-nbd processes on the pod before doing any ungrateful shutdown, yes I remember some of ceph-csi team conversations and it really make sense!

I'm not following.  What happens after all those filesystems get frozen?  Who is going to unfreeze them to clean up user workloads stuck in uninterruptible sleep?

> 
> * Bug in librbd or daemon is not acceptable and should be resolved once we root cause it :-)

But reattach not completing with the 5-minute timeout is?  I guess I'm missing why it seems to be considered "normal work flow".

Reply:
> @idryomov when io-timeout is set to 0, I do not see any difference that is getting affected.
> 
> ```
> [root@linux-vm1 build]# ./bin/rbd-nbd map rbd-pool/image0 --try-netlink --reattach-timeout 300 --io-timeout 0
> /dev/nbd0
> 
> [root@linux-vm1 build]# cat /sys/block/nbd0/queue/io_timeout 
> 30000
> ```
> 
> This is why I treat it to behave the same as default 30 seconds.

Ah, this is not the actual timeout.  This shows how often the timeout handler would fire.  With `io-timeout` set to 0, it should output `Possible stuck request` messages every 30 seconds but never actually time out anything.  Super confusing!

Reply:
> I'm not following. What happens after all those filesystems get frozen? Who is going to unfreeze them to clean up user workloads stuck in uninterruptible sleep?

In a normal workflow the next `attach` should unfreeze the mount. In case if its too late for `attach` to succeed, then the Ceph-CSI Healer should unfreeze the mount points on a recognized exit code from `rbd-nbd attach` failure.

BTW - Ceph-CSI Healer is responsible to do attaching of the rbd-nbd volumes after a pod restart.

> But reattach not completing within the 5-minute timeout is? I guess I'm missing why it seems to be considered "normal work flow".

I get we shouldn't hit 5 mins timeout most of the times, but we cannot expect that to happen all the time for the practical realworld usecases, and so this will be an extra layer of protection. After all it can be an optional feature at rbd-nbd for generic use case.

> > This is why I treat it to behave the same as default 30 seconds.
> 
> Ah, this is not the actual timeout. This shows how often the timeout handler would fire.

Do you mean, its a bug in kernel? You remember we tested this before with various `--io-timeouts` which were actually getting reflected fine at `/sys/block/nbd0/queue/io_timeout `?

Here: https://github.com/ceph/ceph-csi/issues/2204#issuecomment-894033738

> With `io-timeout` set to 0, it should output `Possible stuck request` messages every 30 seconds but never actually time out anything. Super confusing but that seems to be by design ;)

I will have to test it (very soon), but IIUC you are saying if we set  `--io-timeouts` to `0` the nbd driver will consider it and hence should print `Possible stuck request`, but it is probably it is not just getting reflected at `/sys/block/nbd0/queue/io_timeout `?


Thanks @idryomov.

Reply:
Yes.  `queue/io_timeout` shows the value for the timeout handler.  In the timeout handler, the driver can make a choice to abort the request or to ignore the timeout (perhaps initiating some sort of recovery, etc).  Specifying `--io-timeout 0` tells the nbd driver to _not_ abort the request and instead see if it can be restarted on another socket.

> You remember we tested this before with various --io-timeouts which were actually getting reflected fine at `/sys/block/nbd0/queue/io_timeout` ?

A value other than 0 is reflected correctly.  0 is special and doesn't get reflected because it is a bit of a hack in the nbd driver.

Reply:
> In case if its too late for `attach` to succeed, then the Ceph-CSI Healer should unfreeze the mount points on a recognized exit code from `rbd-nbd attach` failure.

Unfreeze without a daemon attached?  What would handle the I/O from the unfreezed filesystem?

Reply:
> Yes. `queue/io_timeout` shows the value for the timeout handler. In the timeout handler, the driver can make a choice to abort the request or to ignore the timeout (perhaps initiating some sort of recovery, etc). Specifying `--io-timeout 0` tells the nbd driver to _not_ abort the request and instead see if it can be restarted on another socket.

Yes, I see this now:

```
[Thu Aug 19 05:12:33 2021] nbd0: detected capacity change from 0 to 2097152
[Thu Aug 19 05:12:47 2021] EXT4-fs (nbd0): mounted filesystem with ordered data mode. Opts: (null). Quota mode: none.
[Thu Aug 19 05:12:58 2021] block nbd0: Receive control failed (result -32)
[Thu Aug 19 05:12:58 2021] block nbd0: Dead connection, failed to find a fallback
[Thu Aug 19 05:13:28 2021] block nbd0: Possible stuck request 00000000e47aef47: control (read@1073676288,4096B). Runtime 30 seconds
[Thu Aug 19 05:13:58 2021] block nbd0: Possible stuck request 00000000e47aef47: control (read@1073676288,4096B). Runtime 60 seconds
[Thu Aug 19 05:14:22 2021] block nbd0: reconnected socket
[Thu Aug 19 05:14:29 2021] block nbd0: Possible stuck request 00000000e47aef47: control (read@1073676288,4096B). Runtime 90 seconds
[Thu Aug 19 05:16:06 2021] block nbd0: NBD_DISCONNECT
[Thu Aug 19 05:16:06 2021] block nbd0: Disconnected due to user request.
[Thu Aug 19 05:16:06 2021] block nbd0: shutting down sockets
```
The `queue/io_timeout` value was quite misleading here.

> 
> > You remember we tested this before with various --io-timeouts which were actually getting reflected fine at `/sys/block/nbd0/queue/io_timeout` ?
> 
> A value other than 0 is reflected correctly. 0 is special and doesn't get reflected because it is a bit of a hack in the nbd driver.

Not an important one, but I wish to fix it someday :-)

> > In case if its too late for `attach` to succeed, then the Ceph-CSI Healer should unfreeze the mount points on a recognized exit code from `rbd-nbd attach` failure.
> 
> Unfreeze without a daemon attached? What would handle the I/O from the unfreezed filesystem?

This will be the same as the case where we set `--io-timeout=0` and cannot attach the process within the `--reattach-timeout`. I expect in both cases the IO should bailout and it's time to map the image freshly i.e. the application pod needs a restart.

That said, probably the next question would be, what are we going to solve by quiescing?

* You guessed it, we guaranty that the data is not corrupted and rbd-nbd is reliable over any pod restarts. If we do not use quiesce some of the filesystem data/metadata might be stuck in the fs/block/kernel layers, in case if the process attach doesn't happen in time, the missing writes might lead to a corrupted backend. On the other hand, by quiescing, we ensure the data will be pushed from fs/block till backend, hence even if the application pod needs a restart, we will be quite confident the data is in a good shape.

Thanks again!

Reply:
Looks like setting `io-timeout` to `0` is what we need. We just want the I/O to be queued until the rbd-nbd process reattaches.

Reply:
@nixpanic Yes, I have updated the PR with the same today.

Reply:
@nixpanic Yes, I have updated the PR with the same today. We also definitely need quiescing for the reason mentioned in the above comment.

Reply:
We would not need `fsfreeze` in case the block-layer does not report errors but just queues the I/O. This should work for both filesystem- and block-mode volumes.

Quiescing takes time (flushing outstanding I/O) and delays restarting the node-plugin Pod. I am not convinced there is a good reason to flushing I/O, as we do expect the pod to restart. Assuming there are issues with restarting, or re-attaching doesn't give me confidence in the solution...

Reply:
Given the scenario where the process is killed(TERM) and brought back to life while the device is held, is a bit special, It will not be a big surprise to arm rbd-nbd with additional safety.

IMHO, at any cost, we shouldn't be in a place that will lead us to corruption. Data loss is secondary to me, which we cannot avoid at the worst.

Reply:
Yesterday there was a discussion around this in the CSI standup meeting, and here is the summary of the discussion:

1. If we do not quiesce on detach, then there is a chance of data corruption, hence we should provide quiesce.
2. For setups where the writes are slow and hence heavily cached on fs (like a few MBs, GBs), the flushing involved as part of fsfreeze might take time and might lead us to hit `reattach-timeout` (note: this can also be reconfigured from current 300sec to higher). Hence we should have the option to disable quiesce on detach at Ceph-CSI.

The pod termination sequence should be improved with the `preHooks`, the hook should ideally wait for `rbd-nbd` processes.

Reply:
> 
> 1. If we do not quiesce on detach, then there is a chance of data corruption, hence we should provide quiesce.
> 

Note that for applications there probably is little difference with or without quiescing. The application does not get informed about the need to flush writes, it can still be interrupted (blocked/paused) while it is writing data.

When updating the DaemonSet for the rbd node-plugin, **all** RBD volumes will start to flush their data when quiescing is done. This will cause a load spike on the Ceph cluster and network.

Reply:
Agreed @nixpanic. This should be almost the same load that takes place if we have to snapshot all the PVs in one go!